BLUESPAWN-client/src/util/configurations/RegistryKey.cpp

```diff
@@ -0,0 +1,485 @@
+	bool RegistryKey::Create(){
+		if(Exists()){
+			return true;
+		}
+
+		if(!hkHive){
+			SetLastError(ERROR_NOT_FOUND);
+			return false;
+		}
+
+		LSTATUS status = RegCreateKeyEx(hkHive, path.c_str(), 0, nullptr, 0, KEY_ALL_ACCESS, nullptr, &hkBackingKey, nullptr);
+		if(status == ERROR_SUCCESS){
+			bKeyExists = true;
+
+			if(_ReferenceCounts.find(hkBackingKey) == _ReferenceCounts.end()){
+				_ReferenceCounts[hkBackingKey] = 1;
+			} else {
+				_ReferenceCounts[hkBackingKey]++;
+			}
+
+			return true;
+		}
+
+		SetLastError(status);
+
+		return false;
+	}
+
+	AllocationWrapper RegistryKey::GetRawValue(const std::wstring& ValueName) const {
+		if(!Exists()){
+			SetLastError(FILE_DOES_NOT_EXIST);
+			return { nullptr, 0 };
+		}
+
+		DWORD dwDataSize{};
+
+		LSTATUS status = RegQueryValueExW(hkBackingKey, ValueName.c_str(), 0, nullptr, nullptr, &dwDataSize);
+		if(status != ERROR_SUCCESS && status != ERROR_MORE_DATA){
+			SetLastError(status);
+			return { nullptr, 0 };
+		}
+
+		auto lpbValue = new BYTE[dwDataSize];
+		status = RegQueryValueExW(hkBackingKey, ValueName.c_str(), 0, nullptr, lpbValue, &dwDataSize);
+		if(status != ERROR_SUCCESS){
+			SetLastError(status);
+			return { nullptr, 0 };
+		}
+
+		return { lpbValue, dwDataSize };
+	}
+
+	std::optional<RegistryType> RegistryKey::GetValueType(const std::wstring& ValueName) const {
+		if(!Exists()){
+			SetLastError(FILE_DOES_NOT_EXIST);
+			return std::nullopt;
+		}
+
+		DWORD dwType{};
+
+		LSTATUS status = RegQueryValueExW(hkBackingKey, ValueName.c_str(), 0, &dwType, nullptr, nullptr);
+		if(status != ERROR_SUCCESS && status != ERROR_MORE_DATA){
+			SetLastError(status);
+			return std::nullopt;
+		}
+
+		if(dwType == REG_SZ){
+			return RegistryType::REG_SZ_T;
+		} else if(dwType == REG_EXPAND_SZ){
+			return RegistryType::REG_EXPAND_SZ_T;
+		} else if(dwType == REG_MULTI_SZ){
+			return RegistryType::REG_MULTI_SZ_T;
+		} else if(dwType == REG_DWORD){
+			return RegistryType::REG_DWORD_T;
+		}
+
+		return RegistryType::REG_BINARY_T;
+	}
+
+	template<class T>
+	std::optional<T> RegistryKey::GetValue(const std::wstring& wsValueName) const {
+		if(ValueExists(wsValueName)){
+			return GetRawValue(wsValueName).Dereference<T>();
+		}
+		return std::nullopt;
+	}
+
+	template std::optional<DWORD> RegistryKey::GetValue(const std::wstring& wsValueName) const;
+
+	template<>
+	std::optional<std::wstring> RegistryKey::GetValue(const std::wstring& wsValueName) const {
+		if(ValueExists(wsValueName)){
+			return GetRawValue(wsValueName).ReadWString();
+		}
+		return std::nullopt;
+	}
+
+	template<>
+	std::optional<std::vector<std::wstring>> RegistryKey::GetValue(const std::wstring& wsValueName) const {
+		if(ValueExists(wsValueName)){
+			std::vector<std::wstring> strings{};
+			std::wstring wsLogString{};
+
+			LPCWSTR data = reinterpret_cast<LPCWSTR>((LPVOID) GetRawValue(wsValueName));
+
+			while(*data){
+				std::wstring str = data;
+				strings.emplace_back(data);
+				wsLogString += str;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'wsLogString' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=4967854)